backend/app/controllers/spree/admin/resource_controller.rb

```diff
@@ -222,7 +223,8 @@ def build_resource
   end
 
   def collection
-    return parent.send(controller_name) if parent?
+    return parent.send(controller_name) if parent? && parent
```

Comment:
Correct me if I'm wrong, but the `&&` parent shouldn't be necessary here, since `parent` will either return a record or throw an exception.

Reply:
I am not sure how else to correct this ... I have followed the code execution to see that the `record_not_found` is thrown, rescued and the redirect is called in the base_controller#resource_not_found as expected.
However, confusingly, the `parent.send(controller_name)` is still called and then the `No_method_error` caused by `undefined method ... for nil:NilClass` occurs again because parent is `nil` in `parent.send(controller_name)`.

This is further baffling since returning the falsy value `nil` (or using [`and return`](https://guides.rubyonrails.org/layouts_and_rendering.html#avoiding-double-render-errors)) after a redirect, as done in the base_controller#resource_not_found, is meant to stop all subsequent actions from executing (it is successfully avoiding "double render errors" already).


The following redirect tests fail as a result of removing `&& parent`:
```
rspec ./spec/controllers/spree/admin/prices_controller_spec.rb:47 # Spree::Admin::PricesController#index existent product id not given cannot find non-existent product
rspec ./spec/controllers/spree/admin/variants_controller_spec.rb:59 # Spree::Admin::VariantsController#index the value of @collection with soft-deleted variants existent product id not given cannot find non-existent product
rspec ./spec/controllers/spree/admin/product_properties_controller_spec.rb:77 # Spree::Admin::ProductPropertiesController#index option values are provided existent product id not given cannot find non-existent product
rspec ./spec/controllers/spree/admin/store_credits_controller_spec.rb:313 # Spree::Admin::StoreCreditsController User does not exist cannot find a store-credit for non-existent user
```

Reply:
@ikraamg have you checked whether `parent` is being overridden anywhere? Perhaps there's an override that still uses `find_by`, causing the method to return `nil`.

Reply:
Out of the above-mentioned controllers causing this `no_method_error`, the variants_controller.rb is the only one with `parent` that is overridden, and this commit has already changed the `find_by` to `find_by!` in the variants controller and then rescues appropriately. 

I have placed breakpoints to follow the execution and in all cases, the `find_by!` does work but after the redirect, the `parent.send(controller_name)` is still called:
> ... I have followed the code execution to see that the `record_not_found` is thrown, rescued and the redirect is called in the base_controller#resource_not_found as expected.
> However, confusingly, the `parent.send(controller_name)` is still called and then the `No_method_error` caused by `undefined method ... for nil:NilClass` occurs again because parent is `nil` in `parent.send(controller_name)`.
> 
> This is further baffling since returning the falsy value `nil` (or using [`and return`](https://guides.rubyonrails.org/layouts_and_rendering.html#avoiding-double-render-errors)) after a redirect, as done in the base_controller#resource_not_found, is meant to stop all subsequent actions from executing (it is successfully avoiding "double render errors" already).

Reply:
Maybe there's another `rescue_from ActiveRecord::RecordNotFound` somewhere?

Reply:
Hi @kennyadsl,  I could not find another `rescue_from ActiveRecord::RecordNotFound` in the controllers or the controllers' parents in which the issue occurs.

Perhaps you may be able to see where the issue lies if I share a step-through of the code below. From what I see, it redirects successfully and then continues executing by attempting to find the `controller_name` which is called by the `return parent.send(controller_name)`, and immediately thereafter the [error](https://user-images.githubusercontent.com/34813339/91338813-c78a4780-e7d5-11ea-9d29-81080b597f13.png) is raised.

```Ruby
Started GET "/admin/products/ruby-5646/prices" for ::1 at 2020-08-26 19:37:29 +0200
   (1.9ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
Processing by Spree::Admin::PricesController#index as HTML
  Parameters: {"product_id"=>"ruby-5646"}
  Spree::User Load (1.1ms)  SELECT "spree_users".* FROM "spree_users" WHERE "spree_users"."deleted_at" IS NULL AND "spree_users"."id" = $1 ORDER BY "spree_users"."id" ASC LIMIT $2  [["id", 1], ["LIMIT", 1]]
  Spree::Role Load (1.6ms)  SELECT "spree_roles".* FROM "spree_roles" INNER JOIN "spree_roles_users" ON "spree_roles"."id" = "spree_roles_users"."role_id" WHERE "spree_roles_users"."user_id" = $1  [["user_id", 1]]
/home/ikraam/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/awesome_nested_set-3.2.1/lib/awesome_nested_set/awesome_nested_set.rb:105: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/home/ikraam/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.2/lib/active_record/associations.rb:1657: warning: The called method `belongs_to' is defined here
/home/ikraam/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/awesome_nested_set-3.2.1/lib/awesome_nested_set/awesome_nested_set.rb:90: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/home/ikraam/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/activerecord-6.0.3.2/lib/active_record/associations.rb:1370: warning: The called method `has_many' is defined here

[227, 236] in /home/ikraam/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/solidus-3dc450477ae5/backend/app/controllers/spree/admin/resource_controller.rb
   227:     end
   228:   end
   229: 
   230:   def collection
   231:     byebug
=> 232:     return parent.send(controller_name) if parent?
   233: 
   234:     if model_class.respond_to?(:accessible_by) && !current_ability.has_block?(params[:action], model_class)
   235:       model_class.accessible_by(current_ability, action)
   236:     else
(byebug) next

[191, 200] in /home/ikraam/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/solidus-3dc450477ae5/backend/app/controllers/spree/admin/resource_controller.rb
   191:     self.class.parent_data
   192:   end
   193: 
   194:   def parent
   195:     byebug
=> 196:     if parent?
   197:       @parent ||= self.class.parent_data[:model_class]
   198:                     .includes(self.class.parent_data[:includes])
   199:                     .find_by!(self.class.parent_data[:find_by] => params["#{parent_model_name}_id"])
   200:       instance_variable_set("@#{parent_model_name}", @parent)
(byebug) next

[192, 201] in /home/ikraam/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/solidus-3dc450477ae5/backend/app/controllers/spree/admin/resource_controller.rb
   192:   end
   193: 
   194:   def parent
   195:     byebug
   196:     if parent?
=> 197:       @parent ||= self.class.parent_data[:model_class]
   198:                     .includes(self.class.parent_data[:includes])
   199:                     .find_by!(self.class.parent_data[:find_by] => params["#{parent_model_name}_id"])
   200:       instance_variable_set("@#{parent_model_name}", @parent)
   201:     else
(byebug) next
  Spree::Product Load (1.7ms)  SELECT "spree_products".* FROM "spree_products" WHERE "spree_products"."deleted_at" IS NULL AND "spree_products"."slug" = $1 LIMIT $2  [["slug", "ruby-5646"], ["LIMIT", 1]]

[202, 211] in /home/ikraam/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/solidus-3dc450477ae5/backend/app/controllers/spree/admin/resource_controller.rb
   202:       Spree::Deprecation.warn "Calling #parent is deprecated on a ResourceController which has not defined a belongs_to"
   203:       nil
   204:     end
   205:   rescue ActiveRecord::RecordNotFound => e
   206:     byebug
=> 207:     resource_not_found(flash_class: e.model.constantize, redirect_url: spree.polymorphic_url([:admin, parent_model_name.pluralize]))
   208:   end
   209: 
   210:   def parent?
   211:     self.class.parent_data.present?
(byebug) next
Redirected to http://localhost:3000/admin/products
Return value is: nil

[65, 74] in /home/ikraam/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/solidus-3dc450477ae5/backend/app/controllers/spree/admin/base_controller.rb
   65: 
   66:       def resource_not_found(flash_class:, redirect_url:)
   67:         flash[:error] = flash_message_for(flash_class.new, :not_found)
   68:         redirect_to redirect_url
   69:         nil
   70:         byebug
=> 71:       end
   72:     end
   73:   end
   74: end
(byebug) step

[136, 145] in /home/ikraam/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/bundler/gems/solidus-3dc450477ae5/backend/app/controllers/spree/admin/resource_controller.rb
   136:     end
   137:   end
   138: 
   139:   def resource_not_found(flash_class: model_class, redirect_url: collection_url)
   140:     super
=> 141:     byebug
   142:   end
   143: 
   144:   def model_class
   145:     "Spree::#{controller_name.classify}".constantize
(byebug) step
Return value is: nil

[139, 148] in /home/ikraam/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.2/lib/action_controller/metal.rb
   139:       false
   140:     end
   141: 
   142:     # Delegates to the class' <tt>controller_name</tt>.
   143:     def controller_name
=> 144:       self.class.controller_name
   145:     end
   146: 
   147:     attr_internal :response, :request
   148:     delegate :session, to: "@_request"
(byebug) next

[19, 28] in /home/ikraam/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/actionpack-6.0.3.2/lib/action_controller/metal/rescue.rb
   19: 
   20:     private
   21:       def process_action(*args)
   22:         super
   23:       rescue Exception => exception
=> 24:         request.env["action_dispatch.show_detailed_exceptions"] ||= show_detailed_exceptions?
   25:         rescue_with_handler(exception) || raise
   26:       end
   27:   end
   28: end
(byebug) next
```

Reply:
Maybe I got it. From [the official rescue_from documentation](https://apidock.com/rails/ActiveSupport/Rescuable/ClassMethods/rescue_from):

> Handlers are inherited. They are searched from right to left, from bottom to top, and up the hierarchy. The handler of the first class for which exception.is_a?(klass) holds true is the one invoked, if any.

This seems to mean (or at least it's what I hope it means 🙂 ) that if you have multiple handlers (methods with the same name) in the hierarchy they will be all executed with that priority described above. 

Could this explain this odd behavior?

Reply:
It seems that `rescue_from` and `rescue` act differently after the redirection.  The `rescue_from` works as expected but the `NoMethodError` is raised when using `rescue` in the method.

The context is lost by removing the `rescue` in the `parent` method and using the `rescue_from`. I have attempted a working solution below, but is there a simple way to find out in the `rescue_from` handler if the error was raised in the `parent` ? 

```ruby
rescue_from ActiveRecord::RecordNotFound do |e|
    if params['controller'].match?(/prices|store_credits|product_properties|adjustments|return_authorizations|customer_returns/)
      resource_not_found(flash_class: e.model.constantize, redirect_url: spree.polymorphic_url([:admin, parent_model_name.pluralize]))
    else
      resource_not_found
    end
  end
```

Reply:
@kennyadsl, @aldesantis I have made an attempt to pass the context into the `rescue_from` by raising a "custom"  `ActiveRecord::RecordNotFound`.

Shall we go with this implementation and remove `&& parent`?

Implementation of rescue_from:
```ruby
rescue_from ActiveRecord::RecordNotFound do |e|
    e.message == 'parent' ? resource_not_found(flash_class: e.model.constantize, redirect_url: spree.polymorphic_url([:admin, parent_model_name.pluralize])) : resource_not_found
end
```
Rescue in parent:
```ruby
rescue ActiveRecord::RecordNotFound => e
    raise ActiveRecord::RecordNotFound.new('parent', e.model, e.primary_key, e.id)
end
```